File: doc/Projects/2020/Project5/DiseaseModeling/project/code/SIRS.py
import matplotlib.pyplot as plt
import numpy as np
import ODESolver

if __name__!="__main__":
    print("SIRS.py imported")

class SIRS(object):
    """Solve the SIRS model as a system of coupled ODEs using the RK4 algorithm"""

    def __init__(self,S,I,R):
        """Class initializer: constructs the state of the system at time t=0.
        Arguments:  S : Initial number of susceptible
                    I : Initial number of infected
                    R : Initial number of recovered
        """
        self.state = [S,I,R]

    def equations(self,state,t):
        """Create the system of ODEs to be used by ODESolver
        """
        S,I,R = state
        N = S+I+R
        for transition in self.transitions:
            exec(transition)

        DS = eval(self.DS_string)
        DI = eval(self.DI_string)
        DR = eval(self.DR_string)
        return [DS,DI,DR]

    def Solve_ODE(self,final_time,dt,terminate=None):
        """Use ODESolver's RungeKutta4 method to solve the system of equations
        Arguments:  final_time : length of the simulation
                    dt : time step
                    terminate (optional) : condition upon which to break the solver loop
        """
        n = round(final_time/dt + 1) # n must be an integer
        print(dt,n)
        t = np.linspace(0,final_time,n)
        ODE = ODESolver.RungeKutta4(self.equations)
        ODE.set_initial_condition(self.state)
        SIR,t = ODE.solve(t,terminate)

        return(SIR,t)

    def Basic_SIRS(self, rStoI,rItoR,rRtoS):
        """Define the equations of the basic SIRS model. This shoul be called
        first after the constructor.
        Arguments:  rStoI : transmission parameter
                    rItoR : recovery parameter
                    rRtoS : immunity loss parameter
        """
        self.a0 = rStoI
        self.b = rItoR
        self.c = rRtoS
        self.time = 0
        self.a = lambda t: self.a0 # Without vital dynamics
        # ROOT-like and ugly. Find a better way to do this.
        self.transitions = ["StoI = self.a(t)*S*I/N",
                            "ItoR = self.b*I",
                            "RtoS = self.c*R"]
        # These will be evaluated in the method "equations"
        self.DS_string = "RtoS - StoI"
        self.DI_string = "StoI - ItoR"
        self.DR_string = "ItoR - RtoS"

    def Vital_dynamics(self,rBirth,rDeath,rIDeath):
        """Add the terms for vital dynamics to the equations
        Arguments:  rBirth : birth rate
                    rDeath : rate of death not related to the disease
                    rIDeath : rate of death caused by the disease
        """
        self.e = rBirth
        self.d = rDeath
        self.dI = rIDeath
        self.DS_string += "+ self.e*(S+I+R) - self.d*S"
        self.DI_string += "- self.d*I - self.dI*I"
        self.DR_string += "- self.d*R"

    def Seasonal_variation(self,amplitude, omega):
        """Change the parameter self.a to account for seasonal variation
        Arguments:  amplitude : maximum deviation from the mean transmission rate
                    omega : angular frequency of the oscillation
        """
        self.A = amplitude
        self.w = omega
        self.a = lambda t: self.A * np.cos(self.w*t) + self.a0

    def Vaccination(self,rate):
        """Add the term for vaccination to the equations
        Argument:   rate(t)  : function of time describing the vaccination rate.
        """
        self.rVacc = lambda t: rate(t)
        self.transitions.append("StoR = self.rVacc(t)*(S>0)")
        self.DS_string += "- StoR"
        self.DR_string += "+ StoR"

File: doc/Projects/2020/Project5/DiseaseModeling/project/code/SIRS.cpp
#include "SIRS.hpp"


/* Constructor: constructs the state of the system at time t=0. Can optionally
set a specific seed for the random number generator.
  Arguments:  int S : Initial number of susceptible
              int I : Initial number of infected
              int R : Initial number of recovered
              int seed (optional) : seed for the random number generator
*/
SIRS::SIRS(int S, int I, int R, int seed/*=1*/){
  cout << "seed = " << seed << endl;
  rng.seed(seed);
  m_state[0] = S;
  m_state[1] = I;
  m_state[2] = R;
  m_state[3] = S+I+R;
}

/* The function calculate_dt calculates the time step such that at most one
individual will move from one group to another. This allowes this number to be
interpreted as a transition probability.
Called by SIRS::Solve_MC
*/
void SIRS::calculate_dt(){
  dt = *min_element(m_invMaxr.begin(),m_invMaxr.end());
  cout << dt << endl;
}

/* The function Solve_MC simulates the system for the given time and writes the
results to the specified file, as tab separated values.
When adding a transition, it should not be necessary to modify this function.
  Arguments:  double final_time : length of the simulation
              int filestep : [NOT IMPLEMENTED] number of timesteps between each write to file
              string outfilename : name of the file to generat
*/
void SIRS::Solve_MC(double final_time, int filestep, string outfilename)
{
  for(auto& x: m_parameters){
    cout << x.first << " : " << x.second << endl;
  }
  calculate_dt();
  int MCs = int(final_time/dt);
  cout << MCs << endl;

  outfile.open(outfilename);
  outfile << m_time << "\t" << m_state[0] << "\t" << m_state[1] << "\t" << m_state[2] << "\n";
  for(int cycle=0;cycle<MCs;cycle++){
    for(int i=0;i<m_functions.size();i++){
      (this->*(m_functions[i]))();
    }
    m_state[3] = m_state[0]+m_state[1]+m_state[2]; // Update N
    m_time += dt;
    outfile << m_time << "\t" << m_state[0] << "\t" << m_state[1] << "\t" << m_state[2] << "\n";
  }
  outfile.close();
}



/*******************************************************************************

The following functions initialize and describe the transitions of the system.
When adding a transition, these two functions should be placed here.
*/


/* The initializer for the basic SIRS model. This should should be called first
after the constructor.
Initialization functions for other transitions should follow this structure.
  Arguments:  double rStoI : transmission parameter
              double rItoR : recovery parameter
              double rRtoS : immunity loss parameter
*/
void SIRS::Basic_SIRS(double rStoI, double rItoR, double rRtoS){
  // Add the parameters to the map m_parameters
  m_parameters["a"]=rStoI;
  m_parameters["b"]=rItoR;
  m_parameters["c"]=rRtoS;
  // Add the address of the function describing the transitions to the vector m_functions
  m_functions.push_back(&SIRS::Basic_SIRS);
  // Add the inverse of the max value of the transition to the vector m_invMaxr
  m_invMaxr.push_back(4/(m_parameters["a"]*m_state[3]));
  m_invMaxr.push_back(1/(m_parameters["b"]*m_state[3]));
  m_invMaxr.push_back(1/(m_parameters["c"]*m_state[3]));
}
/* Function describing the transitions for the basic SIRS model.
Functions describing other transitions should follow this structure.
*/
void SIRS::Basic_SIRS(){
  // Randomly select the transitions according to the transition probabilities
  bool StoI = uniform(rng) < m_parameters["a"]/m_state[3] *m_state[0]*m_state[1]*dt;
  bool ItoR = uniform(rng) < m_parameters["b"]*m_state[1]*dt;
  bool RtoS = uniform(rng) < m_parameters["c"]*m_state[2]*dt;
  // Update the system with the selected transitions
  m_state[0] -= StoI; m_state[1] += StoI;
  m_state[1] -= ItoR; m_state[2] += ItoR;
  m_state[2] -= RtoS; m_state[0] += RtoS;
}



/* The initializer for seasonal variation.
  Arguments:  double amplitude : maximum deviation from the mean transmission rate
              double omega : angular frequency of the oscillation
*/
void SIRS::Seasonal_variation(double amplitude, double omega){
  m_parameters["A"]=amplitude;
  m_parameters["w"]=omega;
  // Replace the Basic_SIRS function by the Seasonal_variation function
  m_functions[0] = (&SIRS::Seasonal_variation);
  m_invMaxr[0] = 4/((m_parameters["a"]+m_parameters["A"])*m_state[3]);
}
/* The transitions for the basic SIRS model, modified with a periodic transmission
parameter.
*/
void SIRS::Seasonal_variation(){
  bool StoI = uniform(rng) < (m_parameters["A"] * cos(m_parameters["w"]*m_time)
                      + m_parameters["a"])/m_state[3] *m_state[0]*m_state[1]*dt;
  bool ItoR = uniform(rng) < m_parameters["b"]*m_state[1]*dt;
  bool RtoS = uniform(rng) < m_parameters["c"]*m_state[2]*dt;
  m_state[0] -= StoI; m_state[1] += StoI;
  m_state[1] -= ItoR; m_state[2] += ItoR;
  m_state[2] -= RtoS; m_state[0] += RtoS;
}



/* The initializer for vital dynamics.
  Arguments:  double rBirth : birth rate
              double rDeath : rate of death not related to the disease
              double rIDeath : rate of death caused by the disease
*/
void SIRS::Vital_dynamics(double rBirth, double rDeath, double rIDeath){
  m_parameters["e"]=rBirth;
  m_parameters["d"]=rDeath;
  m_parameters["dI"]=rIDeath;
  m_functions.push_back(&SIRS::Vital_dynamics);
  m_invMaxr.push_back(1/(m_parameters["e"]*m_state[3]));
  m_invMaxr.push_back(1/(m_parameters["d"]*m_state[3]));
  m_invMaxr.push_back(1/(m_parameters["dI"]*m_state[3]));
}
/* The transitions for vital dynamics. Individuals are added to S by birth, and
removed from all categories by death. Infected individuals can also die from the
disease.
*/
void SIRS::Vital_dynamics(){
  m_state[0] += uniform(rng) < m_parameters["e"]*m_state[3]*dt; // births
  m_state[1] -= uniform(rng) < m_parameters["dI"]*m_state[1]*dt; // deaths from disease
  for(int i=0;i<3;i++){
    m_state[i] -= uniform(rng) < m_parameters["d"]*m_state[i]*dt; // deaths from other causes
  }
}



/* The initializer for vaccination.
  Arguments:  double (*rate)(double,double*) : function describing the vaccination
                rate. Arguments other than time must be passed as a single double pointer.
              double* args : pointer to the arguments other than time to pass to
                the function rate.
*/
void SIRS::Vaccination(double (*rate)(double time, double* args), double* args){
  rVacc = rate;
  m_VaccArgs = args; // Other arguments to pass to the function rate.
  m_functions.push_back(&SIRS::Vaccination);
  // WARNING : Max vaccination rate is not known. If it becomes larger than all
  //         : other transition rates, the algorithm won't be able to keep up.
  //    TODO : Add exception handling for this case.

}
/* The transitions for vacination. Vaccinated individuals move directly from S to R.
*/
void SIRS::Vaccination(){
  // If S=0, there is no one to vaccinate
  bool StoR = uniform(rng) < rVacc(m_time,m_VaccArgs)*(m_state[0]>0)*dt;
  m_state[0] -= StoR;
  m_state[2] += StoR;
}
